src/document.rs

```diff
@@ -542,24 +541,19 @@ impl Document {
         let mut buf = vec![0u8; (length as usize) - 4];
         reader.read_exact(&mut buf)?;
 
-        let mut cursor = std::io::Cursor::new(buf);
-        loop {
-            let tag = cursor.read_u8()?;
+        ensure_read_exactly(buf, "document length longer than contents", |cursor| {
```

Comment:
We can probably move the reading of the bytes into the buffer above into the function too, right?

Reply:
good idea, done